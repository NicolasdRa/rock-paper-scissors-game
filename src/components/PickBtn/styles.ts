import { makeStyles, createStyles, Theme } from "@material-ui/core/styles";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    main: {
      borderRadius: "50%",
      transition: "transform .2s, box-shadow .2s",
      "&:hover": {
        boxShadow: `0px 10px 10px 1px ${theme.palette.common.headerOutline}`,
        transform: "scale(1.04) perspective(0px)",
      },
    },
    btn: {
      padding: "1rem",
      background: `radial-gradient(${theme.palette.common.white}, ${theme.palette.common.score})`,
    },

    btnBig: {
      padding: "1.7rem !important",
    },

    icon: {
      position: "absolute",
      top: "50%",
      left: "50%",
      transform: "translate(-50%, -50%)",
      width: "3rem",
      height: "3rem",
    },

    iconBig: {
      width: "4.5rem",
      height: "4.5rem",
    },

    iconContainer: {
      position: "relative",
      justifyContent: "center",
      alignItems: "center",
      height: "7rem",
      width: "7rem",
      borderRadius: "50%",
      background: `radial-gradient(1rem at center, ${theme.palette.common.score}, ${theme.palette.common.score})`,
      "&::before": {
        content: '""',
        height: "6.6rem",
        width: "7rem",
        backgroundColor: theme.palette.common.white,
        position: "absolute",
        display: "block",
        borderRadius: "50%",
        top: "5%",
      },
    },

    iconContainerBig: {
      height: "10rem",
      width: "10rem",
      "&::before": {
        content: '""',
        height: "10rem",
        width: "10rem",
      },
    },

    btnRock: {
      background: "linear-gradient(hsl(349, 71%, 52%), hsl(349, 70%, 56%))",
      "&::before": {
        content: '""',
        height: "9.5rem",
        width: "9rem",
        backgroundColor: "#a6243c",
        position: "absolute",
        display: "block",
        borderRadius: "50%",
        bottom: "-5%",
        zIndex: -1,
      },
    },

    btnRockBig: {
      "&::before": {
        height: "13rem",
        width: "13rem",
      },
    },

    btnPaper: {
      background: "linear-gradient(hsl(230, 89%, 62%), hsl(230, 89%, 65%))",
      "&::before": {
        content: '""',
        height: "9.5rem",
        width: "9rem",
        backgroundColor: "#2037a8",
        position: "absolute",
        display: "block",
        borderRadius: "50%",
        bottom: "-5%",
        zIndex: -1,
      },
    },

    btnPaperBig: {
      "&::before": {
        height: "13rem",
        width: "13rem",
      },
    },

    btnScissors: {
      background: "linear-gradient(hsl(39, 89%, 49%), hsl(40, 84%, 53%))",
      "&::before": {
        content: '""',
        height: "9.5rem",
        width: "9rem",
        backgroundColor: "#d18d0f",
        position: "absolute",
        display: "block",
        borderRadius: "50%",
        bottom: "-5%",
        zIndex: -1,
      },
    },

    btnScissorsBig: {
      "&::before": {
        height: "13rem",
        width: "13rem",
      },
    },
  })
);
