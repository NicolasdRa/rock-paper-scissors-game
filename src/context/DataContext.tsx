import React, { createContext, useState } from 'react';

export type GlobalContext = {
  score: number;
  setScore: (score: number) => void;
  pick: string;
  setPick: (pick: string) => void;
};

export const DataContext = createContext<GlobalContext>({
  score: 0,
  setScore: () => {},
  pick: '',
  setPick: () => {},
});

interface Props {
  children?: React.ReactNode;
}

export const DataProvider: React.FC<Props> = ({ children }) => {
  const [score, setScore] = useState(12);
  const [pick, setPick] = useState('');

  return (
    <DataContext.Provider value={{ score, setScore, pick, setPick }}>
      {children}
    </DataContext.Provider>
  );
};
