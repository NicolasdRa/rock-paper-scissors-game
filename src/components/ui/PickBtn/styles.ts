import { makeStyles, createStyles, Theme } from "@material-ui/core/styles";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    main: {
      borderRadius: "50%",
      transition: "transform .2s, box-shadow .2s",
      "&:hover": {
        boxShadow: `0px 10px 10px 1px ${theme.palette.common.headerOutline}`,
        transform: "scale(1.05) perspective(0px)",
      },
    },
    btn: {
      padding: "1.1rem",
      background: `radial-gradient(${theme.palette.common.white}, ${theme.palette.common.score})`,

      [theme.breakpoints.down("md")]: {
        padding: ".8rem",
      },
    },

    btnBig: {
      padding: "1.7rem",

      [theme.breakpoints.down("md")]: {
        padding: ".8rem",
      },
    },

    icon: {
      position: "absolute",
      top: "50%",
      left: "50%",
      transform: "translate(-50%, -50%)",
      width: "3rem",
      height: "3rem",

      [theme.breakpoints.down("md")]: {
        width: "2.6rem",
        height: "2.6rem",
      },
    },

    iconBig: {
      width: "4.5rem",
      height: "4.5rem",

      [theme.breakpoints.down("md")]: {
        width: "2.6rem",
        height: "2.6rem",
      },
    },

    iconContainer: {
      position: "relative",
      justifyContent: "center",
      alignItems: "center",
      height: "7rem",
      width: "7rem",
      borderRadius: "50%",
      backgroundColor: theme.palette.common.white,
      boxShadow: `inset 0px 7px 2px ${theme.palette.common.shadow}`,

      [theme.breakpoints.down("md")]: {
        boxShadow: `inset 0px 6px 2px ${theme.palette.common.shadow}`,
        height: "6rem",
        width: "6rem",
      },
    },

    iconContainerBig: {
      height: "10rem",
      width: "10rem",
      "&::before": {
        content: '""',
        height: "10rem",
        width: "10rem",

        [theme.breakpoints.down("md")]: {
          height: "5.8rem",
          width: "6rem",
        },
      },

      [theme.breakpoints.down("md")]: {
        height: "6rem",
        width: "6rem",
      },
    },

    btnRock: {
      background: "linear-gradient(hsl(349, 71%, 52%), hsl(349, 70%, 56%))",
      "&::before": {
        content: '""',
        height: "9.5rem",
        width: "9rem",
        backgroundColor: "#a6243c",
        position: "absolute",
        display: "block",
        borderRadius: "50%",
        bottom: "-5%",
        zIndex: -1,

        [theme.breakpoints.down("md")]: {
          height: "7.5rem",
          width: "7rem",
        },
      },

      [theme.breakpoints.down("md")]: {
        height: "7.3rem",
        width: "7.3rem",
      },
    },

    btnPaper: {
      background: "linear-gradient(hsl(230, 89%, 62%), hsl(230, 89%, 65%))",
      "&::before": {
        content: '""',
        height: "9.5rem",
        width: "9rem",
        backgroundColor: "#2037a8",
        position: "absolute",
        display: "block",
        borderRadius: "50%",
        bottom: "-5%",
        zIndex: -1,

        [theme.breakpoints.down("md")]: {
          height: "7.5rem",
          width: "7rem",
        },
      },

      [theme.breakpoints.down("md")]: {
        height: "7.3rem",
        width: "7.3rem",
      },
    },

    btnScissors: {
      background: "linear-gradient(hsl(39, 89%, 49%), hsl(40, 84%, 53%))",
      "&::before": {
        content: '""',
        height: "9.5rem",
        width: "9rem",
        backgroundColor: "#d18d0f",
        position: "absolute",
        display: "block",
        borderRadius: "50%",
        bottom: "-5%",
        zIndex: -1,

        [theme.breakpoints.down("md")]: {
          height: "7.5rem",
          width: "7rem",
        },
      },

      [theme.breakpoints.down("md")]: {
        height: "7.3rem",
        width: "7.3rem",
      },
    },

    btnRockBig: {
      "&::before": {
        height: "13rem",
        width: "13rem",

        [theme.breakpoints.down("md")]: {
          height: "7.5rem",
          width: "7rem",
        },
      },
    },

    btnPaperBig: {
      "&::before": {
        height: "13rem",
        width: "13rem",

        [theme.breakpoints.down("md")]: {
          height: "7.5rem",
          width: "7rem",
        },
      },
    },

    btnScissorsBig: {
      "&::before": {
        height: "13rem",
        width: "13rem",

        [theme.breakpoints.down("md")]: {
          height: "7.5rem",
          width: "7rem",
        },
      },
    },
  })
);
