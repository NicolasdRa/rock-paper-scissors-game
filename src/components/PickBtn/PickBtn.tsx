import React, { useContext } from "react";
import { Link } from "react-router-dom";
import clsx from "clsx";

import { DataContext } from "../../context/DataContext";
import { Box, IconButton } from "@material-ui/core";
import rockImg from "../../icons/icon-rock.svg";
import paperImg from "../../icons/icon-paper.svg";
import scissorsImg from "../../icons/icon-scissors.svg";

import { useStyles } from "./styles";

interface Props {
  type: string;
  size?: string;
}

export const PickBtn: React.FC<Props> = ({ type, size }) => {
  const classes = useStyles();

  const { setPick } = useContext(DataContext);

  const getImg = (type: string): any => {
    if (type === "rock") {
      return rockImg;
    } else if (type === "paper") {
      return paperImg;
    } else if (type === "scissors") return scissorsImg;
  };

  return (
    <Box className={classes.main}>
      <IconButton
        onClick={() => setPick(type)}
        component={Link}
        to="/game"
        className={clsx(classes.btn, {
          [classes.btnRock]: type === "rock",
          [classes.btnPaper]: type === "paper",
          [classes.btnScissors]: type === "scissors",

          [classes.btnRockBig]: type === "rock" && size === "big",
          [classes.btnPaperBig]: type === "paper" && size === "big",
          [classes.btnScissorsBig]: type === "scissors" && size === "big",
          [classes.btnBig]: size === "big",
        })}
      >
        <Box
          className={clsx(classes.iconContainer, {
            [classes.iconContainerBig]: size === "big",
          })}
        >
          <img
            src={getImg(type)}
            alt="pick btn"
            className={clsx(classes.icon, {
              [classes.iconBig]: size === "big",
            })}
          />
        </Box>
      </IconButton>
    </Box>
  );
};
