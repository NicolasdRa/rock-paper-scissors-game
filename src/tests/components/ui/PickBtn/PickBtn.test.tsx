import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { PickBtn } from "../../../../components/ui/PickBtn/PickBtn";
import { DataContext } from "../../../../context/DataContext";

const contextValue = {
  setScore: jest.fn(),
  score: 0,
  setPick: jest.fn(),
  pick: "",
};

describe("tests on <PickBtn/>", () => {
  const type = "rock";

  test("should render <PickBtn/> correctly", () => {
    const wrapper = mount(
      <DataContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={["/board"]}>
          <PickBtn type={type} />
        </MemoryRouter>
      </DataContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });
  test("should call setPick onClick if not disabled", () => {
    const wrapper = mount(
      <DataContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={["/board"]}>
          <PickBtn type={type} />
        </MemoryRouter>
      </DataContext.Provider>
    );

    wrapper.find("a").prop("onClick")!({} as any);
    expect(contextValue.setPick).toHaveBeenCalledWith("rock");
  });

  test("should not call setPick onClick if disabled", () => {
    const wrapper = mount(
      <DataContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={["/board"]}>
          <PickBtn type={type} disabled />
        </MemoryRouter>
      </DataContext.Provider>
    );

    wrapper.find("a").prop("onClick")!({} as any);
    expect(contextValue.setPick).toHaveBeenCalledTimes(0);
  });
});
