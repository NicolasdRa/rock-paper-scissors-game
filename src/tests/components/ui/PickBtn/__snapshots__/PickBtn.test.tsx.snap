// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <PickBtn/> should render <PickBtn/> correctly 1`] = `
<div
  className="MuiBox-root MuiBox-root-14 makeStyles-main-1"
>
  <a
    aria-disabled={false}
    className="MuiButtonBase-root MuiIconButton-root makeStyles-btn-2 makeStyles-btnRock-8"
    href="/game"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <span
      className="MuiIconButton-label"
    >
      <div
        className="MuiBox-root MuiBox-root-15 makeStyles-iconContainer-6"
      >
        <img
          alt="pick btn"
          className="makeStyles-icon-4"
          src="icon-rock.svg"
        />
      </div>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </a>
</div>
`;
